pinot-segment-local/src/main/java/org/apache/pinot/segment/local/upsert/BaseTableUpsertMetadataManager.java

```diff
@@ -75,6 +99,105 @@ public void init(TableConfig tableConfig, Schema schema, TableDataManager tableD
+      String tableNameWithType = _tableNameWithType;
+      if (Objects.equals(state, "ONLINE")) {
+//        LOGGER.info("Adding or replacing segment: {} for table: {}", segmentName, tableNameWithType);
+
+        // But if table mgr is not created or the segment is not loaded yet, the localMetadata
+        // is set to null. Then, addOrReplaceSegment method will load the segment accordingly.
+        SegmentMetadata localMetadata = getSegmentMetadata(tableDataManager, tableNameWithType, segmentName);
+
+        if (getValidDocIdsSnapshotFile(localMetadata).exists()) {
+          segmentsWithSnapshots.add(segmentName);
+        } else {
+          segmentsWithoutSnapshots.add(segmentName);
+        }
+      }
+    }
+
+    for (String segmentName : segmentsWithSnapshots) {
+      try {
+        loadSegmentFromDisk(tableDataManager, tableConfig, segmentName);
+        _preloadedSegmentsMap.put(segmentName, true);
+      } catch (Exception e) {
+        //TODO: log error
+      }
+    }
+
+    //TODO: Missing step - Change from write only mode to read-write mode
+
+    for (String segmentName : segmentsWithoutSnapshots) {
+      try {
+        loadSegmentFromDisk(tableDataManager, tableConfig, segmentName);
+        _preloadedSegmentsMap.put(segmentName, true);
+      } catch (Exception e) {
+       //TODO: log error
+      }
+    }
+  }
+
+  //TODO: duplicate method from ImmutableSegmentImpl
+  private File getValidDocIdsSnapshotFile(SegmentMetadata segmentMetadata) {
+    return new File(SegmentDirectoryPaths.findSegmentDirectory(segmentMetadata.getIndexDir()),
+        V1Constants.VALID_DOC_IDS_SNAPSHOT_FILE_NAME);
+  }
+
+
+  private void loadSegmentFromDisk(TableDataManager tableDataManager, TableConfig tableConfig, String segmentName)
+      throws Exception {
+    String tableNameWithType = _tableNameWithType;
+    Schema schema = ZKMetadataProvider.getTableSchema(_propertyStore, tableConfig);
+    SegmentZKMetadata zkMetadata =
+        ZKMetadataProvider.getSegmentZKMetadata(_propertyStore, tableNameWithType, segmentName);
+    Preconditions.checkState(zkMetadata != null, "Failed to find ZK metadata for segment: %s, table: %s", segmentName, tableNameWithType);
+
+    // This method might modify the file on disk. Use segment lock to prevent race condition
+    Lock segmentLock = SegmentLocks.getSegmentLock(tableNameWithType, segmentName);
+    try {
+      segmentLock.lock();
+
+      // But if table mgr is not created or the segment is not loaded yet, the localMetadata
+      // is set to null. Then, addOrReplaceSegment method will load the segment accordingly.
+      SegmentMetadata localMetadata = getSegmentMetadata(tableDataManager, tableNameWithType, segmentName);
+      tableDataManager.addOrReplaceSegment(segmentName, new IndexLoadingConfig(tableDataManager.getTableDataManagerConfig().getInstanceDataManagerConfig(), tableConfig, schema),
```

Comment:
the changes of batch loading segments look good to me. 

and fyi, the methods to sync between loading and ready phases

1. the setReady() is supposed to be called at the end of preloadSegments()
2. I'm planning to call untilReady() in most segment mgmt methods on RealtimeTableDataManager, e.g. addSegment(), remoteSegment(), and also this addOrReplaceSegment(). So directly calling it would lead to deadlock. I'll think about either add a boolean on addOrReplaceSegment() or add a new method, w/o waiting on the condition.

```
 @Override
  public void untilReady()
      throws InterruptedException {
    if (_isReady) {
      return;
    }
    _isReadyLock.lock();
    try {
      while (!_isReady) {
        _isReadyCon.await();
      }
    } finally {
      _isReadyLock.unlock();
    }
  }

  private void setIsReady() {
    _isReadyLock.lock();
    try {
      _isReady = true;
      _isReadyCon.signalAll();
    } finally {
      _isReadyLock.unlock();
    }
  }
```

Reply:
I think the RealtimeTableDataManager.addSegment() should be called here to add segment. The addOrReplaceSegment() seems for the Offline table.

Reply:
No, `addOrReplaceSegment` is the only method that's called by HelixInstanceDataManager iiuc.

Internally, it calls the `addSegment` and passed the required params such as segment metadata